In [1]:
import re
from google_play_scraper import app, Sort, reviews
import nltk
import pandas as pd
from nltk.stem import WordNetLemmatizer
from nltk.tokenize import word_tokenize
from num2words import num2words
from nltk.corpus import stopwords
import numpy as np
from textblob import TextBlob
from vaderSentiment.vaderSentiment import SentimentIntensityAnalyzer


# TASK 1

In [2]:
pd.set_option('display.max_rows', None)
pd.set_option('display.max_columns', None)
pd.set_option('display.width', None)
pd.set_option('display.max_colwidth', None)
app_df = pd.read_csv('ImpellerApps.csv')
app_names = list(app_df['app name'])
app_ids = list(app_df['package name'])
df_csv=pd.DataFrame(columns=['Package name','Reviewer name','Review', 'Rating'])
for i in app_ids:
    app_rvws, token = reviews(
            i, # app's ID, found in app's url
            lang='en',            # defaults to 'en'
            country='us',         # defaults to 'us'
            sort=Sort.NEWEST,     # defaults to Sort.MOST_RELEVANT
            count=1000             # defaults to 100
            # , continuation_token=token
        )

    for r in app_rvws:
        row_list={'Package name':i,'Reviewer name':r['userName'],'Review':r['content'], 'Rating':r['score']}
        df_csv = df_csv.append(row_list, ignore_index=True)
        
df_csv.to_csv('eu.e43.impeller.csv')     

In [3]:
app_df1 = pd.read_csv('TwisterApps.csv')
app_names1 = list(app_df1['app name'])
app_ids1 = list(app_df1['package name'])
df_csv1=pd.DataFrame(columns=['Package name','Reviewer name','Review', 'Rating'])
for i in app_ids1:
    app_rvws, token = reviews(
            i, # app's ID, found in app's url
            lang='en',            # defaults to 'en'
            country='us',         # defaults to 'us'
            sort=Sort.NEWEST,     # defaults to Sort.MOST_RELEVANT
            count=1000             # defaults to 100
            # , continuation_token=token
        )
    
    for r in app_rvws:
        row_list={'Package name':i,'Reviewer name':r['userName'],'Review':r['content'], 'Rating':r['score']}
        df_csv1 = df_csv1.append(row_list, ignore_index=True)
        
df_csv1.to_csv('szelok.app.twister.csv')     

In [4]:
app_df1 = pd.read_csv('Fwknop2Apps.csv')
app_names1 = list(app_df1['app name'])
app_ids1 = list(app_df1['package name'])
df_csv1=pd.DataFrame(columns=['Package name','Reviewer name','Review', 'Rating'])
for i in app_ids1:
    app_rvws, token = reviews(
            i, # app's ID, found in app's url
            lang='en',            # defaults to 'en'
            country='us',         # defaults to 'us'
            sort=Sort.NEWEST,     # defaults to Sort.MOST_RELEVANT
            count=1000             # defaults to 100
            # , continuation_token=token
        )
    
    for r in app_rvws:
        row_list={'Package name':i,'Reviewer name':r['userName'],'Review':r['content'], 'Rating':r['score']}
        df_csv1 = df_csv1.append(row_list, ignore_index=True)
        
df_csv1.to_csv('org.cipherdyne.fwknop2.csv') 

In [5]:
app_df1 = pd.read_csv('1x1clockApps.csv')
app_names1 = list(app_df1['app name'])
app_ids1 = list(app_df1['package name'])
df_csv1=pd.DataFrame(columns=['Package name','Reviewer name','Review', 'Rating'])
for i in app_ids1:
    app_rvws, token = reviews(
            i, # app's ID, found in app's url
            lang='en',            # defaults to 'en'
            country='us',         # defaults to 'us'
            sort=Sort.NEWEST,     # defaults to Sort.MOST_RELEVANT
            count=1000             # defaults to 100
            # , continuation_token=token
        )
    
    for r in app_rvws:
        row_list={'Package name':i,'Reviewer name':r['userName'],'Review':r['content'], 'Rating':r['score']}
        df_csv1 = df_csv1.append(row_list, ignore_index=True)
        
df_csv1.to_csv('com.github.wakhub.tinyclock.csv') 

In [6]:
app_df1 = pd.read_csv('ColorClockApps.csv')
app_names1 = list(app_df1['app name'])
app_ids1 = list(app_df1['package name'])
df_csv1=pd.DataFrame(columns=['Package name','Reviewer name','Review', 'Rating'])
for i in app_ids1:
    app_rvws, token = reviews(
            i, # app's ID, found in app's url
            lang='en',            # defaults to 'en'
            country='us',         # defaults to 'us'
            sort=Sort.NEWEST,     # defaults to Sort.MOST_RELEVANT
            count=1000             # defaults to 100
            # , continuation_token=token
        )
    
    for r in app_rvws:
        row_list={'Package name':i,'Reviewer name':r['userName'],'Review':r['content'], 'Rating':r['score']}
        df_csv1 = df_csv1.append(row_list, ignore_index=True)
        
df_csv1.to_csv('com.brianco.colorclock.csv') 

# TASK 2

In [7]:
rev_df = pd.read_csv('eu.e43.impeller.csv')
app_rev = list(rev_df['Review'])
app_id=list(rev_df['Package name'])
 
imp_pre_df=pd.DataFrame(columns=['Package name','Review'])

for r in app_rev:
    app_desc = str(r)
# Punctuations
    # Initialize punctuations string
    punc = r'''!()-[]{};:'"...\,./?@*'''
    # Remove punctuations in description
    for ch in app_desc:
        if ch in punc:
            app_desc = app_desc.replace(ch, "")
# Emojis         
    app_desc_sc = re.sub(r'[^\w\s]', '', app_desc)
    # Emoji pattern
    emoji_pattern = re.compile("["
            u"\U0001F600-\U0001F64F"  # emoticons
            u"\U0001F300-\U0001F5FF"  # symbols & pictographs
            u"\U0001F680-\U0001F6FF"  # transport & map symbols
            u"\U0001F1E0-\U0001F1FF"  # flags (iOS)
                               "]+", flags=re.UNICODE)
    # Remove emojis in description
    emoji_pattern.sub(r'', app_desc_sc)
# #  remove special characters
#     app_desc_sc = re.sub("[^a-zA-Z0-9]+", "",app_desc_sc)
#     re.sub('[^A-Za-z0-9]+ ', '', mystring)
# Number to words
    app_desc_num=app_desc_sc
    # Initialize list of words in description
    word_tokens_num = word_tokenize(app_desc_num)
    # Iterate list and change numbers to words
    for i in word_tokens_num:
        i = re.sub('[^A-Za-z0-9]+', '',i)
#         print(word_tokens_num)
        if i.isdigit():
            app_desc_num = app_desc_num.replace(i, num2words(i))
# Whitespaces    
    app_desc_sp = app_desc_num
    # Remove extra whitespaces using regex string
    app_desc_sp = re.sub(' {2,}', ' ', app_desc_sp)
    
# Lowercase    
    app_desc_low=app_desc_sp
    # Initialize list of words in description
    word_tokens_low = word_tokenize(app_desc_low)
    # Iterate list and change words to lowercase
    for i in word_tokens_low:
        app_desc_low = app_desc_low.replace(i, i.lower())
# Stop words   
    #   Initialize set of english stop words
    stop_words = set(stopwords.words('english'))
    word_tokens_st = word_tokenize(app_desc_low)
    #Initialize list with stop words filtered out and initialize new app description string
    filtered_desc = [w for w in word_tokens_st if not w in stop_words]
    app_desc_stop=' '.join([str(elem) for elem in filtered_desc])
    
# Lemmatize    
    #   Initialize Lemmatizer object 
    wordnet_lemmatizer = WordNetLemmatizer()
    app_desc_lem = app_desc_stop
    word_tokens_lem = word_tokenize(app_desc_lem)
#   Iterate word tokens and Lemmatize words
    for w in word_tokens_lem:
        app_desc_lem = app_desc_lem.replace(w, wordnet_lemmatizer.lemmatize(w))
# Append to dataframe    
    row_list={'Package name':app_id[app_rev.index(r)],'Review':app_desc_lem}
    imp_pre_df = imp_pre_df.append(row_list, ignore_index=True)
    
imp_pre_df.sample(16)

Package name  \
1194      com.twitter.android   
1255      com.twitter.android   
2780      com.facebook.katana   
1776      com.twitter.android   
927       com.twitter.android   
1909      com.facebook.katana   
2081      com.facebook.katana   
1485      com.twitter.android   
871   com.keylesspalace.tusky   
1788      com.twitter.android   
1723      com.twitter.android   
1834      com.twitter.android   
2625  com.keylesspalace.tusky   
1493      com.twitter.android   
1282      com.twitter.android   
2749      com.facebook.katana   

                                                                                                                      Review  
1194                                                              good social network unfortunately since new update problem  
1255                                                                                                                  superb  
2780                                                                         fb easiest social medium app understand useimho  
1776                                                                                                            good app use  
927                                                                                                      awesome application  
1909                                                                                                             old picture  
2081                                                                                                                    stat  
1485                                                                                                               good apps  
871                                                                                        would good app instance could log  
1788                                                                                                     app crashed twitter  
1723                                                                                                               fantastic  
1834                                                                               retrieve tweet past tweet gone please fix  
2625                                                                                                                    nice  
1493                                                                                           installed told cant installed  
1282  issue app click onto tweet profile soon click tweet entire page reset youre back top profile get really annoying fixed  
2749                                                                                                fine app geef three star

In [8]:
rev_df = pd.read_csv('szelok.app.twister.csv')
app_rev = list(rev_df['Review'])
app_id=list(rev_df['Package name'])
 
twi_pre_df=pd.DataFrame(columns=['Package name','Review'])

for r in app_rev:
    app_desc = str(r)
# Punctuations
    # Initialize punctuations string
    punc = r'''!()-[]{};:'"...\,./?@*'''
    # Remove punctuations in description
    for ch in app_desc:
        if ch in punc:
            app_desc = app_desc.replace(ch, "")
# Emojis         
    app_desc_sc = re.sub(r'[^\w\s]', '', app_desc)
    # Emoji pattern
    emoji_pattern = re.compile("["
            u"\U0001F600-\U0001F64F"  # emoticons
            u"\U0001F300-\U0001F5FF"  # symbols & pictographs
            u"\U0001F680-\U0001F6FF"  # transport & map symbols
            u"\U0001F1E0-\U0001F1FF"  # flags (iOS)
                               "]+", flags=re.UNICODE)
    # Remove emojis in description
    emoji_pattern.sub(r'', app_desc_sc)
# Number to words
    app_desc_num=app_desc_sc
    # Initialize list of words in description
    word_tokens_num = word_tokenize(app_desc_num)
    # Iterate list and change numbers to words
    for i in word_tokens_num:
        if i.isdigit():
            app_desc_num = app_desc_num.replace(i, num2words(i))
# Whitespaces    
    app_desc_sp = app_desc_num
    # Remove extra whitespaces using regex string
    app_desc_sp = re.sub(' {2,}', ' ', app_desc_sp)
    
# Lowercase    
    app_desc_low=app_desc_sp
    # Initialize list of words in description
    word_tokens_low = word_tokenize(app_desc_low)
    # Iterate list and change words to lowercase
    for i in word_tokens_low:
        app_desc_low = app_desc_low.replace(i, i.lower())
# Stop words   
    #   Initialize set of english stop words
    stop_words = set(stopwords.words('english'))
    word_tokens_st = word_tokenize(app_desc_low)
    #Initialize list with stop words filtered out and initialize new app description string
    filtered_desc = [w for w in word_tokens_st if not w in stop_words]
    app_desc_stop=' '.join([str(elem) for elem in filtered_desc])
    
# Lemmatize    
    #   Initialize Lemmatizer object 
    wordnet_lemmatizer = WordNetLemmatizer()
    app_desc_lem = app_desc_stop
    word_tokens_lem = word_tokenize(app_desc_lem)
#   Iterate word tokens and Lemmatize words
    for w in word_tokens_lem:
        app_desc_lem = app_desc_lem.replace(w, wordnet_lemmatizer.lemmatize(w))
# Append to dataframe    
    row_list={'Package name':app_id[app_rev.index(r)],'Review':app_desc_lem}
    twi_pre_df = twi_pre_df.append(row_list, ignore_index=True)
    
twi_pre_df.sample(16)

Package name  \
847  io.starnetwork.app   
948  io.starnetwork.app   
661  io.starnetwork.app   
632  io.starnetwork.app   
542  io.starnetwork.app   
74   io.starnetwork.app   
439  io.starnetwork.app   
289  io.starnetwork.app   
356  io.starnetwork.app   
723  io.starnetwork.app   
239  io.starnetwork.app   
666  io.starnetwork.app   
181  io.starnetwork.app   
870  io.starnetwork.app   
258  io.starnetwork.app   
221  io.starnetwork.app   

                                                                                                                                                                                                                                                                                                                                                       Review  
847                                                                                                                                                                                                                                                                                                                                             lie kyc token  
948                                                                                                                                                                                                                                                                                                                                                        gg  
661                                                                                                                                                                                                                                                                                                                                                  good app  
632                                                                                                                                                                                                                                                                      disappointed app say release kyc hit 500k didnt happen make fun people rollette draw  
542                                                                                                                                                                                                                                                                                                                                                      best  
74                                                                                                                                                                                                                                                              invite code ruezzz great way earn passive crypto add invite code join team invite code ruezzz  
439  nice inform star network refferal code lighta gamified application wiil value dont huge value future keep mining could make u billionaire day see cryptocurrency thai going well day day member mut seriou free cryptocurrency bee network give u freecryptocurrency collect bee coin mining also collect bee reffering friend big revolution generation  
289                                                                                                                                                                                                                                                                                                                                       referal code lidman  
356                                                                                                                                                                                                                                                      invitation code markgpelesco wow rate five star cause good platform application easy used similar pi  
723                            

In [9]:
rev_df = pd.read_csv('org.cipherdyne.fwknop2.csv')
app_rev = list(rev_df['Review'])
app_id=list(rev_df['Package name'])
 
fwk_pre_df=pd.DataFrame(columns=['Package name','Review'])

for r in app_rev:
    app_desc = str(r)
# Punctuations
    # Initialize punctuations string
    punc = r'''!()-[]{};:'"...\,./?@*'''
    # Remove punctuations in description
    for ch in app_desc:
        if ch in punc:
            app_desc = app_desc.replace(ch, "")
# Emojis         
    app_desc_sc = re.sub(r'[^\w\s]', '', app_desc)
    # Emoji pattern
    emoji_pattern = re.compile("["
            u"\U0001F600-\U0001F64F"  # emoticons
            u"\U0001F300-\U0001F5FF"  # symbols & pictographs
            u"\U0001F680-\U0001F6FF"  # transport & map symbols
            u"\U0001F1E0-\U0001F1FF"  # flags (iOS)
                               "]+", flags=re.UNICODE)
    # Remove emojis in description
    emoji_pattern.sub(r'', app_desc_sc)
# Number to words
    app_desc_num=app_desc_sc
    # Initialize list of words in description
    word_tokens_num = word_tokenize(app_desc_num)
    # Iterate list and change numbers to words
    for i in word_tokens_num:
        if i.isdigit():
            app_desc_num = app_desc_num.replace(i, num2words(i))
# Whitespaces    
    app_desc_sp = app_desc_num
    # Remove extra whitespaces using regex string
    app_desc_sp = re.sub(' {2,}', ' ', app_desc_sp)
    
# Lowercase    
    app_desc_low=app_desc_sp
    # Initialize list of words in description
    word_tokens_low = word_tokenize(app_desc_low)
    # Iterate list and change words to lowercase
    for i in word_tokens_low:
        app_desc_low = app_desc_low.replace(i, i.lower())
# Stop words   
    #   Initialize set of english stop words
    stop_words = set(stopwords.words('english'))
    word_tokens_st = word_tokenize(app_desc_low)
    #Initialize list with stop words filtered out and initialize new app description string
    filtered_desc = [w for w in word_tokens_st if not w in stop_words]
    app_desc_stop=' '.join([str(elem) for elem in filtered_desc])
    
# Lemmatize    
    #   Initialize Lemmatizer object 
    wordnet_lemmatizer = WordNetLemmatizer()
    app_desc_lem = app_desc_stop
    word_tokens_lem = word_tokenize(app_desc_lem)
#   Iterate word tokens and Lemmatize words
    for w in word_tokens_lem:
        app_desc_lem = app_desc_lem.replace(w, wordnet_lemmatizer.lemmatize(w))
# Append to dataframe    
    row_list={'Package name':app_id[app_rev.index(r)],'Review':app_desc_lem}
    fwk_pre_df = fwk_pre_df.append(row_list, ignore_index=True)
    
fwk_pre_df.sample(16)

Package name  \
43     com.sonelli.portknocker   
35     com.sonelli.portknocker   
100       me.impa.knockonports   
78   com.xargsgrep.portknocker   
17     com.smpaine.portknocker   
13     com.smpaine.portknocker   
99        me.impa.knockonports   
3       org.cipherdyne.fwknop2   
55   com.xargsgrep.portknocker   
84        me.impa.knockonports   
85        me.impa.knockonports   
89        me.impa.knockonports   
18     com.smpaine.portknocker   
9       org.cipherdyne.fwknop2   
19     com.smpaine.portknocker   
94        me.impa.knockonports   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     Review  
43                                                                                                                                                                                                                                                                                                                                                                                                                     app slows s5 tremendously importantly knock sequence received server  
35                                                                                                                                                                                                                                                                                                                                                                                                                                           work perfectly thanks go developer handle tool  
100                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     say  
78                                                                                                                                                                                                                                                                                                                                                                                                                                                           work like charm galaxy tab two  
17                                                                                                                                                                                                                                                                                                                                                                                                                                                         please make integration juicessh  
13                                                                                                                                                                                                                                                                                                                                                                                                                                                          please add importexport setting  
99                                                                                           

In [10]:
rev_df = pd.read_csv('com.github.wakhub.tinyclock.csv')
app_rev = list(rev_df['Review'])
app_id=list(rev_df['Package name'])
 
tc_pre_df=pd.DataFrame(columns=['Package name','Review'])

for r in app_rev:
    app_desc = str(r)
# Punctuations
    # Initialize punctuations string
    punc = r'''!()-[]{};:'"...\,./?@*'''
    # Remove punctuations in description
    for ch in app_desc:
        if ch in punc:
            app_desc = app_desc.replace(ch, "")
# Emojis         
    app_desc_sc = re.sub(r'[^\w\s]', '', app_desc)
    # Emoji pattern
    emoji_pattern = re.compile("["
            u"\U0001F600-\U0001F64F"  # emoticons
            u"\U0001F300-\U0001F5FF"  # symbols & pictographs
            u"\U0001F680-\U0001F6FF"  # transport & map symbols
            u"\U0001F1E0-\U0001F1FF"  # flags (iOS)
                               "]+", flags=re.UNICODE)
    # Remove emojis in description
    emoji_pattern.sub(r'', app_desc_sc)
# Number to words
    app_desc_num=app_desc_sc
    # Initialize list of words in description
    word_tokens_num = word_tokenize(app_desc_num)
    # Iterate list and change numbers to words
    for i in word_tokens_num:
        if i.isdigit():
            app_desc_num = app_desc_num.replace(i, num2words(i))
# Whitespaces    
    app_desc_sp = app_desc_num
    # Remove extra whitespaces using regex string
    app_desc_sp = re.sub(' {2,}', ' ', app_desc_sp)
    
# Lowercase    
    app_desc_low=app_desc_sp
    # Initialize list of words in description
    word_tokens_low = word_tokenize(app_desc_low)
    # Iterate list and change words to lowercase
    for i in word_tokens_low:
        app_desc_low = app_desc_low.replace(i, i.lower())
# Stop words   
    #   Initialize set of english stop words
    stop_words = set(stopwords.words('english'))
    word_tokens_st = word_tokenize(app_desc_low)
    #Initialize list with stop words filtered out and initialize new app description string
    filtered_desc = [w for w in word_tokens_st if not w in stop_words]
    app_desc_stop=' '.join([str(elem) for elem in filtered_desc])
    
# Lemmatize    
    #   Initialize Lemmatizer object 
    wordnet_lemmatizer = WordNetLemmatizer()
    app_desc_lem = app_desc_stop
    word_tokens_lem = word_tokenize(app_desc_lem)
#   Iterate word tokens and Lemmatize words
    for w in word_tokens_lem:
        app_desc_lem = app_desc_lem.replace(w, wordnet_lemmatizer.lemmatize(w))
# Append to dataframe    
    row_list={'Package name':app_id[app_rev.index(r)],'Review':app_desc_lem}
    tc_pre_df = tc_pre_df.append(row_list, ignore_index=True)
    
tc_pre_df.sample(16)

Package name  \
1318  com.chibatching.worldclockwidget   
2055  com.chibatching.worldclockwidget   
187        com.nullium.nicesimpleclock   
2099  com.chibatching.worldclockwidget   
1960  com.chibatching.worldclockwidget   
2068  com.chibatching.worldclockwidget   
1664  com.chibatching.worldclockwidget   
385        com.nullium.nicesimpleclock   
1912  com.chibatching.worldclockwidget   
1042       com.nullium.nicesimpleclock   
1375  com.chibatching.worldclockwidget   
244        com.nullium.nicesimpleclock   
1677  com.chibatching.worldclockwidget   
806        com.nullium.nicesimpleclock   
478        com.nullium.nicesimpleclock   
1545  com.chibatching.worldclockwidget   

                                                                                                                                                       Review  
1318                                                                              clock freeze everytime need open update time instead automatically adjusted  
2055                                                                                                                                                  perfect  
187                                                                                                                                                      nice  
2099                                                                                                                         us air push put banner ad screen  
1960                                                                                                                           cool simple simple always best  
2068                                                                                                                                    poor apptime updating  
1664                  clock lag behind one plus 7t device requires manual refresh show correct time didnt see problem moto g5 plus device compatibility issue  
385   excellent finally dateclock widget provides 1x1 1x2 widget thanks much dev donate using would get full mark could hide time display date please include  
1912                                                                                                        simple elegant useful wish would found app sooner  
1042                                                                                                              great app great response time producer also  
1375                                                                                                 search city unreliable doesnt canberra capital australia  
244                                                        good balance feature simplicity lately however time hasnt updating sometimes open app setting wake  
1677                                                                                                                        work pretty well thanks great app  
806                                                                                                                                      perfect widget thank  
478                                                                              gripe clock stick occasionally previous time cant trust present current time  
1545                                                                                                                 widget clock didnt refresh automatically

In [11]:
rev_df = pd.read_csv('com.brianco.colorclock.csv')
app_rev = list(rev_df['Review'])
app_id=list(rev_df['Package name'])
cc_pre_df=pd.DataFrame(columns=['Package name','Review'])

for r in app_rev:
    app_desc = str(r)
# Punctuations
    # Initialize punctuations string
    punc = r'''!()-[]{};:'"...\,./?@*'''
    # Remove punctuations in description
    for ch in app_desc:
        if ch in punc:
            app_desc = app_desc.replace(ch, "")
# Emojis         
    app_desc_sc = re.sub(r'[^\w\s]', '', app_desc)
    # Emoji pattern
    emoji_pattern = re.compile("["
            u"\U0001F600-\U0001F64F"  # emoticons
            u"\U0001F300-\U0001F5FF"  # symbols & pictographs
            u"\U0001F680-\U0001F6FF"  # transport & map symbols
            u"\U0001F1E0-\U0001F1FF"  # flags (iOS)
                               "]+", flags=re.UNICODE)
    # Remove emojis in description
    emoji_pattern.sub(r'', app_desc_sc)
# Number to words
    app_desc_num=app_desc_sc
    # Initialize list of words in description
    word_tokens_num = word_tokenize(app_desc_num)
    # Iterate list and change numbers to words
    for i in word_tokens_num:
        if i.isdigit():
            app_desc_num = app_desc_num.replace(i, num2words(i))
# Whitespaces    
    app_desc_sp = app_desc_num
    # Remove extra whitespaces using regex string
    app_desc_sp = re.sub(' {2,}', ' ', app_desc_sp)
    
# Lowercase    
    app_desc_low=app_desc_sp
    # Initialize list of words in description
    word_tokens_low = word_tokenize(app_desc_low)
    # Iterate list and change words to lowercase
    for i in word_tokens_low:
        app_desc_low = app_desc_low.replace(i, i.lower())
# Stop words   
    #   Initialize set of english stop words
    stop_words = set(stopwords.words('english'))
    word_tokens_st = word_tokenize(app_desc_low)
    #Initialize list with stop words filtered out and initialize new app description string
    filtered_desc = [w for w in word_tokens_st if not w in stop_words]
    app_desc_stop=' '.join([str(elem) for elem in filtered_desc])
    
# Lemmatize    
    #   Initialize Lemmatizer object 
    wordnet_lemmatizer = WordNetLemmatizer()
    app_desc_lem = app_desc_stop
    word_tokens_lem = word_tokenize(app_desc_lem)
#   Iterate word tokens and Lemmatize words
    for w in word_tokens_lem:
        app_desc_lem = app_desc_lem.replace(w, wordnet_lemmatizer.lemmatize(w))
# Append to dataframe    
    row_list={'Package name':app_id[app_rev.index(r)],'Review':app_desc_lem}
    cc_pre_df = cc_pre_df.append(row_list, ignore_index=True)
    
cc_pre_df.sample(16)

Package name  \
134                        org.ngosite.android.myclock   
116                        org.ngosite.android.myclock   
142                                 com.peerem.android   
3                               com.brianco.colorclock   
66   photo.effecttech.digital.clock.digitalclockwidget   
38   photo.effecttech.digital.clock.digitalclockwidget   
47   photo.effecttech.digital.clock.digitalclockwidget   
129                        org.ngosite.android.myclock   
40   photo.effecttech.digital.clock.digitalclockwidget   
122                        org.ngosite.android.myclock   
22   photo.effecttech.digital.clock.digitalclockwidget   
98                         org.ngosite.android.myclock   
20   photo.effecttech.digital.clock.digitalclockwidget   
57   photo.effecttech.digital.clock.digitalclockwidget   
126                        org.ngosite.android.myclock   
30   photo.effecttech.digital.clock.digitalclockwidget   

                                                                                                                                                                                                                                                                                                 Review  
134                                                                                                                                                                                                                introduced helen love ur apps look forward innovative well real user friendly apps u  
116  после установки занимает five hundred twelve кб плюсы бесплатно нет рекламы платежей лишних разрешений показывается день недели минусы ограниченная демоверсия платного приложения нет выбора языка включая русский формата отображения даты разные иконки приложения на маркете и после установки  
142                                                                                                                                                                                                                                               recent update ad longer appearing every message thank  
3                                                                                                                                                                                                                                                                     needed clock complement wallpaper  
66                                                                                                                                                                                                                                                                       app sometimes wont update time  
38                                                                                                                                                                                                                                                                                     widget beautiful  
47                                                                                                                                                                                                                                                                                                  gud  
129                                                                                                                                                                                                                                                                                      exactly needed  
40                                                                                                                                                                                                                                                                  great app one kind clock widget try  
122                                                                        

# Bonus of TASK 2

I believe the choice of stop words one uses has quite a sizeable impact on the analysis of reviews. This is because reviews come from users of apps who exist within a specific domain or scope so certain words would be more common among the sets of reviews from those users. Thus if one can identify stopwords that are distinct to each app's domain then they can extend the list of stopwords to include the ones they have identified. This seems useful to me because as a result, the pre-processing algorithm used will return an even lighter data set so there are fewer unimportant words passed to whatever topic model one decides to use. Similarly one could replace stopwords that aren't exactly applicable in the given domain in order to retain important information that would otherwise be lost.

The risk of not using a customized stopwords list is first, losing important data that could drastically impact your analysis and secondly, spending more time than you should on the analysis due to having words in the pre-processed reviews that would be considered stopwords from the standpoint of that specific domain.

# TASK 3

## Textblob Sentiment Analysis

In [12]:
# impeller
imp_app_rev = list(imp_pre_df['Review'])
imp_app_id=list(imp_pre_df['Package name'])
# print(imp_pre_df.head(5))
# twister
twi_app_rev = list(twi_pre_df['Review'])
twi_app_id=list(twi_pre_df['Package name'])
twi_pre_df.head(5)
# fwknop2
fwk_app_rev = list(fwk_pre_df['Review'])
fwk_app_id=list(fwk_pre_df['Package name'])
# 1x1 clock
tc_app_rev = list(tc_pre_df['Review'])
tc_app_id=list(tc_pre_df['Package name'])
# color clock
cc_app_rev = list(cc_pre_df['Review'])
cc_app_id=list(cc_pre_df['Package name'])

imp_sen_df1=pd.DataFrame(columns=['Package name','Review','Polarity'])
twi_sen_df1=pd.DataFrame(columns=['Package name','Review','Polarity'])
fwk_sen_df1=pd.DataFrame(columns=['Package name','Review','Polarity'])
tc_sen_df1=pd.DataFrame(columns=['Package name','Review','Polarity'])
cc_sen_df1=pd.DataFrame(columns=['Package name','Review','Polarity'])

for st in imp_app_rev:
    
    b = TextBlob(str(st))
    row_list={'Package name':imp_app_id[imp_app_rev.index(st)],'Review':str(st),'Polarity':b.sentiment[0]}
#     print(row_list)
    imp_sen_df1 = imp_sen_df1.append(row_list, ignore_index=True)


for st in twi_app_rev:
    
    b = TextBlob(str(st))
    row_list={'Package name':twi_app_id[twi_app_rev.index(st)],'Review':str(st),'Polarity':b.sentiment[0]}
    twi_sen_df1 = twi_sen_df1.append(row_list, ignore_index=True)
# print(twi_sen_df1.head(5))


for st in fwk_app_rev:
#     print(st)
    b = TextBlob(str(st))
    row_list={'Package name':fwk_app_id[fwk_app_rev.index(st)],'Review':str(st),'Polarity':b.sentiment[0]}
    fwk_sen_df1 = fwk_sen_df1.append(row_list, ignore_index=True)


for st in tc_app_rev:
    b = TextBlob(str(st))
    row_list={'Package name':tc_app_id[tc_app_rev.index(st)],'Review':str(st),'Polarity':b.sentiment[0]}
    tc_sen_df1 = tc_sen_df1.append(row_list, ignore_index=True)


for st in cc_app_rev:
    b = TextBlob(str(st))
    row_list={'Package name':cc_app_id[cc_app_rev.index(st)],'Review':str(st),'Polarity':b.sentiment[0]}
    cc_sen_df1 = cc_sen_df1.append(row_list, ignore_index=True)


frames = [imp_sen_df1, twi_sen_df1, fwk_sen_df1, tc_sen_df1, cc_sen_df1]

result = pd.concat(frames)
result.sample(250)

Package name  \
1809                                com.twitter.android   
117                             com.keylesspalace.tusky   
1931                   com.chibatching.worldclockwidget   
2121                   com.chibatching.worldclockwidget   
2371                                com.facebook.katana   
754                                  io.starnetwork.app   
224                             com.keylesspalace.tusky   
879                                  io.starnetwork.app   
365                                  io.starnetwork.app   
1167              com.primedev.digitalclock.widget.free   
1752                   com.chibatching.worldclockwidget   
30                              com.avrapps.clockwidget   
2142                   com.chibatching.worldclockwidget   
804                                  io.starnetwork.app   
1433                   com.chibatching.worldclockwidget   
126                         org.ngosite.android.myclock   
249                         com.nullium.nicesimpleclock   
1306                                com.twitter.android   
2047                   com.chibatching.worldclockwidget   
1843                   com.chibatching.worldclockwidget   
1731                                com.twitter.android   
1680                                com.twitter.android   
699                             com.keylesspalace.tusky   
48                              com.sonelli.portknocker   
190                         com.nullium.nicesimpleclock   
2392                                com.facebook.katana   
698                                  io.starnetwork.app   
957                                  io.starnetwork.app   
228                             com.keylesspalace.tusky   
1089              com.primedev.digitalclock.widget.free   
92                              com.keylesspalace.tusky   
431                         com.nullium.nicesimpleclock   
2865                                com.facebook.katana   
2163                            com.keylesspalace.tusky   
376                                  io.starnetwork.app   
1183              com.primedev.digitalclock.widget.free   
2586                            com.keylesspalace.tusky   
7                                 com.abarts.widgetforu   
1191                        com.nullium.nicesimpleclock   
1169              com.primedev.digitalclock.widget.free   
1169                                com.twitter.android   
530                                  io.starnetwork.app   
192                                  io.starnetwork.app   
1131              com.primedev.digitalclock.widget.free   
1716                        com.nullium.nicesimpleclock   
34                              com.keylesspalace.tusky   
354                             com.keylesspalace.tusky   
2066                   com.chibatching.worldclockwidget   
2                                com.brianco.colorclock   
1646                   com.chibatching.worldclockwidget   
1505                   com.chibatching.worldclockwidget   
2177                   com.chibatching.worldclockwidget   
827                         com.nullium.nicesimpleclock   
338                         com.nullium.nicesimpleclock   
1792                   com.chibatching.worldclockwidget   
505                             com.keylesspalace.tusky   
1408                                com.twitter.android   
380                             com.keylesspalace.tusky   
1413                   com.chibatching.worldclockwidget   
520                                  io.starnetwork.app   
1960                                com.twitter.android   
849                             com.keylesspalace.tusky   
1440                   com.chibatching.worldclockwidget   
955                                  io.starnetwork.app   
1375                   com.chibatching.worldclockwidget   
3                                com.brianco.colorclock   
2819                            com.keylesspalace.tusky   
2301                           

## Vader Sentiment Analysis

In [13]:
imp_sen_df2=pd.DataFrame(columns=['Package name','Review','Polarity'])
twi_sen_df2=pd.DataFrame(columns=['Package name','Review','Polarity'])
fwk_sen_df2=pd.DataFrame(columns=['Package name','Review','Polarity'])
tc_sen_df2=pd.DataFrame(columns=['Package name','Review','Polarity'])
cc_sen_df2=pd.DataFrame(columns=['Package name','Review','Polarity'])
analyzer = SentimentIntensityAnalyzer()
    
for st in imp_app_rev:
    sentiment_dict = analyzer.polarity_scores(str(st))
    row_list={'Package name':imp_app_id[imp_app_rev.index(st)],'Review':str(st),'Polarity':sentiment_dict['compound']}
    imp_sen_df2 = imp_sen_df2.append(row_list, ignore_index=True)


for st in twi_app_rev:
    
    sentiment_dict = analyzer.polarity_scores(str(st))
    row_list={'Package name':twi_app_id[twi_app_rev.index(st)],'Review':str(st),'Polarity':sentiment_dict['compound']}
    twi_sen_df2 = twi_sen_df2.append(row_list, ignore_index=True)


for st in fwk_app_rev:
    sentiment_dict = analyzer.polarity_scores(str(st))
    row_list={'Package name':fwk_app_id[fwk_app_rev.index(st)],'Review':str(st),'Polarity':sentiment_dict['compound']}
    fwk_sen_df2 = fwk_sen_df2.append(row_list, ignore_index=True)


for st in tc_app_rev:
    sentiment_dict = analyzer.polarity_scores(str(st))
    row_list={'Package name':tc_app_id[tc_app_rev.index(st)],'Review':str(st),'Polarity':sentiment_dict['compound']}
    tc_sen_df2 = tc_sen_df2.append(row_list, ignore_index=True)


for st in cc_app_rev:
    sentiment_dict = analyzer.polarity_scores(str(st))
    row_list={'Package name':cc_app_id[cc_app_rev.index(st)],'Review':str(st),'Polarity':sentiment_dict['compound']}
    cc_sen_df2 = cc_sen_df2.append(row_list, ignore_index=True)
imp_sen_df2.sample(250)

frames2 = [imp_sen_df2, twi_sen_df2, fwk_sen_df2, tc_sen_df2, cc_sen_df2]

result2 = pd.concat(frames2)
result2.sample(250)

Package name  \
417                             com.keylesspalace.tusky   
805                             com.keylesspalace.tusky   
2055                        com.nullium.nicesimpleclock   
326                                  io.starnetwork.app   
1884                   com.chibatching.worldclockwidget   
83                              com.keylesspalace.tusky   
240                                  io.starnetwork.app   
75    photo.effecttech.digital.clock.digitalclockwidget   
174                             com.keylesspalace.tusky   
1867                                com.twitter.android   
1452                        com.nullium.nicesimpleclock   
506                             com.keylesspalace.tusky   
978                                 com.twitter.android   
2642                                com.facebook.katana   
1315                   com.chibatching.worldclockwidget   
18                                com.abarts.widgetforu   
291                         com.nullium.nicesimpleclock   
26                                   io.starnetwork.app   
564                                  io.starnetwork.app   
193                         com.nullium.nicesimpleclock   
1139                                com.twitter.android   
528                                  io.starnetwork.app   
348                             com.keylesspalace.tusky   
1949                        com.nullium.nicesimpleclock   
571                             com.keylesspalace.tusky   
2258                                com.facebook.katana   
1881                                com.twitter.android   
2671                                com.facebook.katana   
1849                   com.chibatching.worldclockwidget   
236                         com.nullium.nicesimpleclock   
823                             com.keylesspalace.tusky   
2613                                com.facebook.katana   
837                         com.nullium.nicesimpleclock   
2102                        com.nullium.nicesimpleclock   
2656                                com.facebook.katana   
1289                                com.twitter.android   
1826                                com.twitter.android   
770                                  io.starnetwork.app   
482                             com.keylesspalace.tusky   
402                                  io.starnetwork.app   
122                         org.ngosite.android.myclock   
626                         com.nullium.nicesimpleclock   
525                         com.nullium.nicesimpleclock   
1237                   com.chibatching.worldclockwidget   
334                                  io.starnetwork.app   
497                         com.nullium.nicesimpleclock   
1202                        com.nullium.nicesimpleclock   
380                                  io.starnetwork.app   
1726                                com.twitter.android   
33    photo.effecttech.digital.clock.digitalclockwidget   
2527                            com.keylesspalace.tusky   
855                             com.keylesspalace.tusky   
779                         com.nullium.nicesimpleclock   
1914                   com.chibatching.worldclockwidget   
856                                  io.starnetwork.app   
224                         com.nullium.nicesimpleclock   
565                         com.nullium.nicesimpleclock   
2006                            com.keylesspalace.tusky   
1707                                com.twitter.android   
2230                            com.keylesspalace.tusky   
798                             com.keylesspalace.tusky   
530                             com.keylesspalace.tusky   
1825                                com.twitter.android   
1644                   com.chibatching.worldclockwidget   
281                         com.nullium.nicesimpleclock   
884                         com.nullium.nicesimpleclock   
2432                                com.facebook.katana   
861                            

## TASK 3 Part V

In [14]:
imp_com_df=pd.DataFrame(columns=['Package name','Review','Textblob Polarity', 'Vader Polarity'])
imp_com_df['Package name']=list(imp_sen_df1['Package name'])
imp_com_df['Review']=list(imp_sen_df1['Review'])
imp_com_df['Textblob Polarity']=list(imp_sen_df1['Polarity'])
imp_com_df['Vader Polarity']=list(imp_sen_df2['Polarity'])
imp_com_df.sample(250)

Package name  \
1688      com.twitter.android   
1902      com.facebook.katana   
925       com.twitter.android   
2491      com.twitter.android   
592   com.keylesspalace.tusky   
2100  com.keylesspalace.tusky   
2673      com.facebook.katana   
1432      com.twitter.android   
2061      com.twitter.android   
1002      com.twitter.android   
949       com.twitter.android   
1517      com.twitter.android   
248   com.keylesspalace.tusky   
2373      com.facebook.katana   
997   com.keylesspalace.tusky   
367   com.keylesspalace.tusky   
1477      com.twitter.android   
664   com.keylesspalace.tusky   
27          org.andstatus.app   
2861  com.keylesspalace.tusky   
960       com.twitter.android   
2360      com.facebook.katana   
2397      com.facebook.katana   
2650      com.facebook.katana   
261   com.keylesspalace.tusky   
2716      com.facebook.katana   
2624  com.keylesspalace.tusky   
1537      com.twitter.android   
1656      com.twitter.android   
2239      com.facebook.katana   
268   com.keylesspalace.tusky   
2039  com.keylesspalace.tusky   
621   com.keylesspalace.tusky   
2118      com.facebook.katana   
2000  com.keylesspalace.tusky   
1655      com.twitter.android   
955       com.twitter.android   
705   com.keylesspalace.tusky   
813   com.keylesspalace.tusky   
468   com.keylesspalace.tusky   
1592      com.twitter.android   
1685      com.twitter.android   
908   com.keylesspalace.tusky   
2127      com.facebook.katana   
1093      com.twitter.android   
1138      com.twitter.android   
2272      com.facebook.katana   
2128      com.facebook.katana   
1797      com.twitter.android   
120   com.keylesspalace.tusky   
722   com.keylesspalace.tusky   
2550      com.facebook.katana   
1291      com.twitter.android   
2601      com.facebook.katana   
300   com.keylesspalace.tusky   
2052      com.facebook.katana   
1558      com.twitter.android   
578   com.keylesspalace.tusky   
449   com.keylesspalace.tusky   
21          org.andstatus.app   
2565      com.facebook.katana   
2836      com.facebook.katana   
77    com.keylesspalace.tusky   
25          org.andstatus.app   
1112      com.twitter.android   
607   com.keylesspalace.tusky   
1349      com.twitter.android   
149   com.keylesspalace.tusky   
2347  com.keylesspalace.tusky   
838   com.keylesspalace.tusky   
2236      com.facebook.katana   
789   com.keylesspalace.tusky   
2755      com.facebook.katana   
2232  com.keylesspalace.tusky   
1266      com.twitter.android   
2215      com.facebook.katana   
603   com.keylesspalace.tusky   
2553      com.twitter.android   
1388      com.twitter.android   
1174      com.twitter.android   
338   com.keylesspalace.tusky   
377   com.keylesspalace.tusky   
662   com.keylesspalace.tusky   
1090      com.twitter.android   
1372  com.keylesspalace.tusky   
1550      com.twitter.android   
2010      com.twitter.android   
497   com.keylesspalace.tusky   
1131      com.twitter.android   
1924      com.facebook.katana   
2333      com.facebook.katana   
1078      com.twitter.android   
134   com.keylesspalace.tusky   
1970      com.facebook.katana   
1726      com.twitter.android   
2773      com.facebook.katana   
2456      com.facebook.katana   
153   com.keylesspalace.tusky   
2151      com.facebook.katana   
506   com.keylesspalace.tusky   
2289      com.facebook.katana   
78    com.keylesspalace.tusky   
2413      com.facebook.katana   
1148      com.twitter.android   
933       com.twitter.android   
36    com.keylesspalace.tusky   
2470      com.twitter.android   
1088      com.twitter.android   
303   com.keylesspalace.tusky   
2054      com.facebook.katana   
411   com.keylesspalace.tusky   
1709      com.twitter.android   
2694      com.facebook.katana   
1767      com.twitter.android   
1588      com.twitter.android   
2529  com.keylesspalace.tusky   
417   com.keylesspalace.tusky   
1984      com.facebook.katana   
1608  com.keylesspalace.tusky   
924   com.keylesspalace.tusky   
1296      com.twitter.an

The sentiments retrieved by Textblob and Vader seem fairly similar for some reviews but upon closer manual inspection with the Columns side by side as shown above I noticed that there are some reviews I have identified as negative such as "app contains malware created antihuman right sjw cultist avoid cost" from line 591 for which Textblob sentiment analysis outputs a polarity of 0.285714 which is overall a positive sentiment meanwhile, Vader sentiment analysis outputs a compound value of -0.0516 which is overall a negative sentiment and is a more accurate analysis of the review. My interpretation of the similarity is that some reviews use distinctively positive or negative language so both libraries would output overall positive or negative sentiments. However, the difference can be seen when the review doesn't exactly use distinct negative words such as slang and domain-specific words in this case Vader seems to perform a better analysis of the text. 
I think the best option for review analysis of my apps would be Vader sentiment analysis because the type of language app reviewers are most likely to use is going to be very informal and contain slang and terminology outside everyday English.

# TASK 4

In [60]:
import gensim
import gensim.corpora as corpora
from pprint import pprint
from gensim.similarities import MatrixSimilarity
from gensim.matutils import cossim
import numpy as np

def sent_to_words(sentences):
    for sentence in sentences:
        yield(gensim.utils.simple_preprocess(str(sentence),
                                             deacc=True))
              
tc_list=tc_sen_df2.loc[tc_sen_df2['Package name'] == 'com.github.wakhub.tinyclock']['Review'].values.tolist()
tc_words = list(sent_to_words(tc_list))
# Create Dictionary
id2word = corpora.Dictionary(tc_words)
# Create Corpus
texts = tc_words
# Term Document Frequency
corpus = [id2word.doc2bow(text) for text in texts]

# number of topics
num_topics = 10
# Build LDA model
lda_model = gensim.models.LdaMulticore(corpus=corpus,
                                       id2word=id2word,
                                       num_topics=num_topics
                                      )
# Print the Keyword in the 10 topics
pprint(lda_model.print_topics())

[(0,
  '0.038*"please" + 0.038*"size" + 0.038*"reduce" + 0.038*"useful" + '
  '0.038*"thanks" + 0.038*"font" + 0.038*"advice" + 0.003*"rubbish" + '
  '0.003*"nan" + 0.003*"say"'),
 (1,
  '0.042*"say" + 0.042*"nan" + 0.030*"color" + 0.026*"square" + 0.022*"time" + '
  '0.022*"fit" + 0.022*"widget" + 0.020*"selection" + 0.020*"longer" + '
  '0.020*"press"'),
 (2,
  '0.033*"time" + 0.025*"square" + 0.024*"color" + 0.023*"colour" + '
  '0.022*"click" + 0.021*"display" + 0.019*"phone" + 0.017*"one" + '
  '0.016*"make" + 0.016*"ok"'),
 (3,
  '0.061*"perfect" + 0.031*"font" + 0.031*"would" + 0.031*"size" + '
  '0.031*"simple" + 0.031*"job" + 0.016*"color" + 0.016*"ninety" + '
  '0.016*"last" + 0.016*"clock"'),
 (4,
  '0.070*"clock" + 0.036*"widget" + 0.026*"extra" + 0.025*"awesome" + '
  '0.024*"tap" + 0.024*"stock" + 0.024*"app" + 0.024*"alarm" + 0.024*"open" + '
  '0.014*"fit"'),
 (5,
  '0.034*"color" + 0.027*"background" + 0.025*"many" + 0.022*"one" + '
  '0.018*"alarm" + 0.018*"setting" +

## TASK 4 Part III

The review topics are very closely related to the software features we manually extracted in assignment 2. They are also very similar to the automated n-grams we extracted in assignment 3. We can take the 4th and 5th topics from the list above, for example, the words clock (excluding app because it appears in topic 1) and time have the most contributions to those topics and one of the features extracted in assignment 2 was 1x1 clock, in assignment 3 the bi-grams extracted were small 1x1, 1x1 clock, and clock widget and the tri-grams were small 1x1 clock and 1x1 clock widget. These extracted n-grams are also very similar to the topics extracted above.

In [62]:
avr_list=tc_sen_df2.loc[tc_sen_df2['Package name'] == 'com.avrapps.clockwidget']['Review'].values.tolist()
avr_words = list(sent_to_words(avr_list))
# Create Dictionary
id2word = corpora.Dictionary(avr_words)
# Create Corpus
texts = avr_words
# Term Document Frequency
corpus_a = [id2word.doc2bow(text) for text in texts]

# number of topics
num_topics_c = 5
# Build LDA model
lda_model_avr = gensim.models.LdaMulticore(corpus=corpus_a,
                                       id2word=id2word,
                                       num_topics=num_topics_c)
# Print the Keyword in the 5 topics
pprint(lda_model_avr.print_topics())


[(0,
  '0.032*"super" + 0.025*"app" + 0.018*"touch" + 0.018*"opened" + 0.017*"show" '
  '+ 0.017*"day" + 0.017*"even" + 0.016*"work" + 0.016*"ad" + 0.016*"cant"'),
 (1,
  '0.438*"good" + 0.018*"super" + 0.017*"need" + 0.016*"size" + 0.014*"date" + '
  '0.013*"work" + 0.013*"day" + 0.012*"doesnt" + 0.011*"font" + 0.011*"year"'),
 (2,
  '0.070*"super" + 0.035*"work" + 0.035*"time" + 0.032*"doesnt" + '
  '0.032*"clock" + 0.024*"good" + 0.017*"digital" + 0.017*"open" + '
  '0.017*"tapping" + 0.017*"look"'),
 (3,
  '0.054*"work" + 0.048*"doesnt" + 0.028*"five" + 0.023*"time" + 0.021*"font" '
  '+ 0.021*"app" + 0.020*"note" + 0.020*"would" + 0.020*"simple" + '
  '0.020*"clean"'),
 (4,
  '0.058*"work" + 0.041*"good" + 0.041*"app" + 0.040*"doesnt" + 0.040*"im" + '
  '0.023*"date" + 0.023*"nice" + 0.022*"simple" + 0.022*"view" + '
  '0.022*"something"')]


In [54]:
# tc_list=tc_sen_df2.loc[tc_sen_df2['Package name'] == 'com.github.wakhub.tinyclock']['Review'].values.tolist()

nice_list=tc_sen_df2.loc[tc_sen_df2['Package name'] == 'com.nullium.nicesimpleclock']['Review'].values.tolist()


nice_words = list(sent_to_words(nice_list))
# Create Dictionary
id2word = corpora.Dictionary(nice_words)
# Create Corpus
texts = nice_words
# Term Document Frequency
corpus_n = [id2word.doc2bow(text) for text in texts]

# number of topics
num_topics_c = 5
# Build LDA model
lda_model_nice = gensim.models.LdaMulticore(corpus=corpus_n,
                                       id2word=id2word,
                                       num_topics=num_topics_c)
# Print the Keyword in the 5 topics
pprint(lda_model_nice.print_topics())



[(0,
  '0.033*"clock" + 0.031*"time" + 0.026*"simple" + 0.019*"app" + '
  '0.015*"widget" + 0.012*"one" + 0.012*"doesnt" + 0.011*"nice" + '
  '0.011*"update" + 0.010*"perfect"'),
 (1,
  '0.039*"clock" + 0.034*"app" + 0.023*"time" + 0.017*"great" + 0.014*"best" + '
  '0.012*"widget" + 0.011*"good" + 0.011*"option" + 0.009*"need" + '
  '0.008*"date"'),
 (2,
  '0.029*"time" + 0.029*"nice" + 0.018*"clock" + 0.017*"app" + 0.014*"good" + '
  '0.011*"excellent" + 0.009*"one" + 0.009*"well" + 0.009*"widget" + '
  '0.007*"like"'),
 (3,
  '0.049*"app" + 0.020*"nice" + 0.015*"widget" + 0.015*"simple" + 0.014*"love" '
  '+ 0.013*"clock" + 0.013*"time" + 0.011*"great" + 0.010*"option" + '
  '0.010*"would"'),
 (4,
  '0.043*"time" + 0.036*"widget" + 0.022*"clock" + 0.018*"app" + 0.017*"great" '
  '+ 0.016*"simple" + 0.014*"date" + 0.010*"like" + 0.010*"easy" + '
  '0.009*"set"')]


In [52]:
# tc_list=tc_sen_df2.loc[tc_sen_df2['Package name'] == 'com.github.wakhub.tinyclock']['Review'].values.tolist()

prime_list=tc_sen_df2.loc[tc_sen_df2['Package name'] == 'com.primedev.digitalclock.widget.free']['Review'].values.tolist()


prime_words = list(sent_to_words(prime_list))
# Create Dictionary
id2word = corpora.Dictionary(prime_words)
# Create Corpus
texts = prime_words
# Term Document Frequency
corpus_p = [id2word.doc2bow(text) for text in texts]
# number of topics
num_topics_c = 5
# Build LDA model
lda_model_p = gensim.models.LdaMulticore(corpus=corpus_p,
                                       id2word=id2word,
                                       num_topics=num_topics_c
                                      )
# Print the Keyword in the 5 topics
pprint(lda_model_p.print_topics())


[(0,
  '0.030*"app" + 0.026*"clock" + 0.025*"widget" + 0.018*"working" + '
  '0.016*"would" + 0.014*"pro" + 0.011*"style" + 0.011*"great" + '
  '0.011*"attractive" + 0.010*"like"'),
 (1,
  '0.031*"clock" + 0.031*"app" + 0.022*"version" + 0.020*"screen" + '
  '0.020*"doesnt" + 0.017*"time" + 0.017*"work" + 0.017*"font" + 0.014*"look" '
  '+ 0.012*"style"'),
 (2,
  '0.043*"good" + 0.024*"nice" + 0.019*"option" + 0.018*"clock" + 0.015*"work" '
  '+ 0.015*"color" + 0.014*"app" + 0.011*"use" + 0.010*"thanks" + '
  '0.010*"look"'),
 (3,
  '0.047*"app" + 0.043*"clock" + 0.023*"widget" + 0.016*"really" + '
  '0.016*"nice" + 0.015*"time" + 0.013*"font" + 0.013*"work" + 0.013*"great" + '
  '0.011*"option"'),
 (4,
  '0.022*"app" + 0.020*"time" + 0.019*"color" + 0.017*"premium" + 0.016*"work" '
  '+ 0.016*"version" + 0.014*"please" + 0.014*"easy" + 0.013*"remove" + '
  '0.012*"set"')]


In [53]:
# tc_list=tc_sen_df2.loc[tc_sen_df2['Package name'] == 'com.github.wakhub.tinyclock']['Review'].values.tolist()

sim_list=tc_sen_df2.loc[tc_sen_df2['Package name'] == 'com.chibatching.worldclockwidget']['Review'].values.tolist()


sim_words = list(sent_to_words(sim_list))
# Create Dictionary
id2word = corpora.Dictionary(sim_words)
# Create Corpus
texts = sim_words
# Term Document Frequency
corpus_s = [id2word.doc2bow(text) for text in texts]
# number of topics
num_topics_c = 5
# Build LDA model
lda_model_s = gensim.models.LdaMulticore(corpus=corpus_s,
                                       id2word=id2word,
                                       num_topics=num_topics_c)
# Print the Keyword in the 5 topics
# matrix = np.zeros(shape=(len(lda_model_avr[corpus]), len(lda_model_avr[corpus])),dtype=float)
pprint(lda_model_s.print_topics())


[(0,
  '0.068*"time" + 0.037*"widget" + 0.034*"app" + 0.032*"update" + '
  '0.022*"doesnt" + 0.015*"simple" + 0.013*"good" + 0.009*"zone" + '
  '0.009*"automatically" + 0.009*"work"'),
 (1,
  '0.047*"time" + 0.020*"clock" + 0.020*"widget" + 0.017*"app" + 0.014*"one" + '
  '0.014*"simple" + 0.013*"use" + 0.010*"easy" + 0.009*"update" + '
  '0.009*"need"'),
 (2,
  '0.057*"time" + 0.029*"doesnt" + 0.016*"app" + 0.015*"work" + 0.015*"great" '
  '+ 0.014*"update" + 0.012*"refresh" + 0.010*"clock" + 0.008*"simple" + '
  '0.008*"screen"'),
 (3,
  '0.067*"time" + 0.019*"app" + 0.013*"widget" + 0.011*"dont" + 0.011*"good" + '
  '0.011*"simple" + 0.011*"update" + 0.011*"clock" + 0.010*"zone" + '
  '0.009*"doesnt"'),
 (4,
  '0.040*"app" + 0.026*"widget" + 0.022*"time" + 0.021*"clock" + 0.013*"work" '
  '+ 0.012*"simple" + 0.010*"set" + 0.010*"update" + 0.009*"nice" + '
  '0.009*"great"')]


## TASK 4 Part V  

## 1x1 clock Vs AVR-Apps-Digital Clock Widget

In [68]:
hits=0
losses=0

for i in range(0, len(lda_model[corpus])):
    for j in range(0, len(lda_model_avr[corpus_a])):
        doc_1 = lda_model[corpus][i]
        doc_2 = lda_model_avr[corpus_a][j]
        if (cossim(doc_1, doc_2)>=0.5):
            hits+=1
            if(i<5 and j<10):
                print('Cosine_Similarity_'+str(i)+'_'+str(j)+':',cossim(doc_1, doc_2)) 
        else:
            losses+=1
print('')
print("Hits:", hits)
print("Losses:", losses)
print("Hit Ratio:",str('{0:.2f}'.format((hits/(hits+losses)*100)))+'%')

#         print('Cosine_Similarity_'+str(i)+'_'+str(j)+':',cossim(doc_1, doc_2))
       
    

Cosine_Similarity_0_2: 0.9983913994818281
Cosine_Similarity_0_7: 0.9969894984351599
Cosine_Similarity_0_9: 1.000000043457619
Cosine_Similarity_1_0: 0.9939909704417249
Cosine_Similarity_3_2: 0.9973884367560737
Cosine_Similarity_3_9: 0.9953479117771877
Cosine_Similarity_4_1: 0.9477206670428889
Cosine_Similarity_4_3: 0.9920686483931952

Hits: 728
Losses: 3952
Hit Ratio: 15.56%


## 1x1 clock Vs Nice Simple Clock (Widget)

In [69]:
hits=0
losses=0

for i in range(0, len(lda_model[corpus])):
    for j in range(0, len(lda_model_nice[corpus_n])):
        doc_1 = lda_model[corpus][i]
        doc_2 = lda_model_nice[corpus_n][j]
        if (cossim(doc_1, doc_2)>=0.5):
            hits+=1
            if(i<5 and j<10):
                print('Cosine_Similarity_'+str(i)+'_'+str(j)+':',cossim(doc_1, doc_2)) 
        else:
            losses+=1
print('')
print("Hits:", hits)
print("Losses:", losses)
print("Hit Ratio:",str('{0:.2f}'.format((hits/(hits+losses)*100)))+'%')

Cosine_Similarity_0_2: 0.9996377129004417
Cosine_Similarity_0_4: 0.9968607336386613
Cosine_Similarity_0_8: 0.9920918682634291
Cosine_Similarity_1_3: 0.9936898735073274
Cosine_Similarity_3_1: 0.9902929469931722
Cosine_Similarity_3_8: 0.9955379830502843

Hits: 3929
Losses: 35599
Hit Ratio: 9.94%


## 1x1 clock Vs Prime Dev Studio- Digital Clock Widget

In [70]:
hits=0
losses=0

for i in range(0, len(lda_model[corpus])):
    for j in range(0, len(lda_model_p[corpus_p])):
        doc_1 = lda_model[corpus][i]
        doc_2 = lda_model_p[corpus_p][j]
        if (cossim(doc_1, doc_2)>=0.5):
            hits+=1
            if(i<5 and j<10):
                print('Cosine_Similarity_'+str(i)+'_'+str(j)+':',cossim(doc_1, doc_2))
        else:
            losses+=1
print('')
print("Hits:", hits)
print("Losses:", losses)
print("Hit Ratio:",str('{0:.2f}'.format((hits/(hits+losses)*100)))+'%')

Cosine_Similarity_1_0: 0.9944283064455681
Cosine_Similarity_1_2: 0.9944261100338854
Cosine_Similarity_1_5: 0.994281987196484
Cosine_Similarity_1_6: 0.9944275996168026
Cosine_Similarity_1_8: 0.9942767495145796
Cosine_Similarity_1_9: 0.994428332557902
Cosine_Similarity_4_3: 0.9479174163779077

Hits: 362
Losses: 3202
Hit Ratio: 10.16%


## 1x1 clock Vs Prime Simple World Clock Widget

In [71]:
hits=0
losses=0

for i in range(0, len(lda_model[corpus])):
    for j in range(0, len(lda_model_s[corpus_s])):
        doc_1 = lda_model[corpus][i]
        doc_2 = lda_model_s[corpus_s][j]
        if (cossim(doc_1, doc_2)>=0.5):
            hits+=1
            if(i<5 and j<10):
                print('Cosine_Similarity_'+str(i)+'_'+str(j)+':',cossim(doc_1, doc_2))
        else:
            losses+=1
print('')
print("Hits:", hits)
print("Losses:", losses)
print("Hit Ratio:",str('{0:.2f}'.format((hits/(hits+losses)*100)))+'%')

Cosine_Similarity_0_8: 0.9920994861222514
Cosine_Similarity_1_3: 0.9902670641825342
Cosine_Similarity_1_4: 0.9944286861116632
Cosine_Similarity_1_7: 0.9940026344778496
Cosine_Similarity_1_9: 0.9939776817066399
Cosine_Similarity_3_0: 0.9973121679801433
Cosine_Similarity_3_5: 0.9955160917177739

Hits: 2714
Losses: 26842
Hit Ratio: 9.18%


I mainly used Cosine Similarity to compare the topics between my Assigned app and the chosen similar and competitor apps. Then I used the criteria of a cosine similarity of over 0.5 to identify reviews that share similar topics and calculated hits and losses as well as the percentage of hits to the total reviews.

# Bonus of TASK 4

In [80]:
def format_topics_sentences(ldamodel=lda_model, corpus=corpus, texts=tc_list):
    # Init output
    sent_topics_df = pd.DataFrame()

    # Get main topic in each document
    for i, row in enumerate(ldamodel[corpus]):
        row = sorted(row, key=lambda x: (x[1]), reverse=True)
        # Get the Dominant topic, Perc Contribution and Keywords for each document
        for j, (topic_num, prop_topic) in enumerate(row):
            if j == 0:  # => dominant topic
                wp = ldamodel.show_topic(topic_num)
                topic_keywords = ", ".join([word for word, prop in wp])
                sent_topics_df = sent_topics_df.append(pd.Series([int(topic_num), round(prop_topic,4), topic_keywords]), ignore_index=True)
            else:
                break
    sent_topics_df.columns = ['Dominant_Topic', 'Percent_Contribution', 'Topic_Keywords']

    # Add original text to the end of the output
    contents = pd.Series(texts)
    sent_topics_df = pd.concat([sent_topics_df, contents], axis=1)
    return(sent_topics_df)


df_topic_sents_keywords = format_topics_sentences(ldamodel=lda_model, corpus=corpus, texts=tc_list)

# Format
df_dominant_topic = df_topic_sents_keywords.reset_index(drop=True)
df_dominant_topic.columns = ['Dominant_Topic', 'Topic_Percent_Contribution', 'Keywords', 'Review']

# Show
df_dominant_topic.head(36)
# print(len(lda_model[corpus]))

Dominant_Topic  Topic_Percent_Contribution  \
0              1.0                      0.9795   
1              3.0                      0.7000   
2              6.0                      0.9667   
3              4.0                      0.7750   
4              2.0                      0.9743   
5              8.0                      0.8875   
6              3.0                      0.9550   
7              6.0                      0.7000   
8              6.0                      0.9357   
9              5.0                      0.9571   
10             4.0                      0.9526   
11             3.0                      0.9250   
12             8.0                      0.9437   
13             3.0                      0.9182   
14             7.0                      0.9550   
15             8.0                      0.9100   
16             8.0                      0.9000   
17             7.0                      0.8714   
18             3.0                      0.7750   
19             6.0                      0.9667   
20             1.0                      0.5500   
21             9.0                      0.5500   
22             5.0                      0.8875   
23             0.0                      0.8875   
24             4.0                      0.9000   
25             1.0                      0.5500   
26             8.0                      0.5500   
27             5.0                      0.9308   
28             4.0                      0.9750   
29             1.0                      0.5500   
30             9.0                      0.5500   
31             8.0                      0.5500   
32             9.0                      0.5500   
33             1.0                      0.5500   
34             1.0                      0.5500   
35             9.0                      0.5500   

                                                                             Keywords  \
0                say, nan, color, square, time, fit, widget, selection, longer, press   
1                 perfect, font, would, size, simple, job, color, ninety, last, clock   
2                app, one, color, font, fifteen, get, hundred, background, nice, look   
3                    clock, widget, extra, awesome, tap, stock, app, alarm, open, fit   
4                   time, square, color, colour, click, display, phone, one, make, ok   
5                   time, love, use, system, display, work, size, cause, font, widget   
6                 perfect, font, would, size, simple, job, color, ninety, last, clock   
7                app, one, color, font, fifteen, get, hundred, background, nice, look   
8                app, one, color, font, fifteen, get, hundred, background, nice, look   
9                 color, background, many, one, alarm, setting, tap, time, five, date   
10                   clock, widget, extra, awesome, tap, stock, app, alarm, open, fit   
11                perfect, font, would, size, simple, job, color, ninety, last, clock   
12                  time, love, use, system, display, work, size, cause, font, widget   
13                perfect, font, would, size, simple, job, color, ninety, last, clock   
14  colour, rubbish, background, put, perfect, pic, code, applicable, listing, trying   
15                  time, love, use, system, display, work, size, cause, font, widget   
16                  time, love, use, system, display, work, size, cause, font, widget   
17  colour, rubbish, background, put, perfect, pic, code, applicable, listing, trying   
18                perfect, font, would, size, simple, job, color, ninety, last, clock   
19               app, one, color, font, fifteen, get, hundred, background, nice, look   
20               say, nan, color, square, time, fit, widget, selection, longer, press   
21         rubbish, phone, time, display, click, colour, iv, expensive, dont, chinese   
22                color, background, many, one, alarm, setting, tap, time, five, date   
23              p

## Bonus of TASK 4 Part II

### High Priority Reviews
1. REV-0: "It was a learning curve. The one square cut off the p.m. touch and hold the 
    square, it will be highlighted and then you can pull down on it to make it 
    longer. That was too long for me so I put it back up to the one square and 
    then I pulled it sideways to make it longer, that was better. If you're 
    going to change the color of the text or background, you put the word of 
    the color you want, press ok, bring up the change color selection again and 
    than press ok again, then your selection will be entered." by James Stutzman. 


2. REV-2: "I'm disappointed in this app. It looks nice on here, but it doesn't look 
    nice on my phone. I have no idea how in the hell this app got 4 stars. It's 
    fairly worthless. How do you pick colors for anything without an 
    explanation of what code is for what color??? This app is STUPID!!!!!!!!!!" by Amber Adams.


3. REV-3: "Time does not fit into the widget" by Elle H.


4. REV-5: "Terrible. Display cuts of part of the time shown. Deleted." by A Google user.


5. REV-11: "Virtually perfect. Font size adjustment, and directly launching the clock app would top it off." by Rob Saunders.


6. REV-14: "How to know the colour codes aabbffgg what is this? I am trying to put background colours but only 2  3 colours can b applicable in pics they are showing more colours...how to get them?are there any colour listings?" by Siddharth Shah ss.


7. REV-10: "When I tap into Clock App, they're many apps for me to select, but without my Clock itself. Please include as it should be the obvious app people want to associate with thid widget." by Liau Kok Meng.


8. REV-15: "Wanted to display UTC, but it's linked to my system time, and it changed my system time." by Joel Marion.


9. REV-19: "The time font is too big and gets cut off. So instead of reading 11:15, I get 1:15. Which sucks. Other than that the app is solid. Please advise on changing font size and I'll rate it again." by Matt.


10. REV-27: "5 if you remove time and date settings . no one ever uses them. Straight to alarms on tap please." by neil murray.

My Thinking for the high priority reviews selected above that these reviews contain either bugs informally reported by customers or general ideas for improving the user experience. For example, if you take a loost at the first review REV-0 the review mainly talks about the shape of the widget anf how it the borders cut off the time in the pm so it seems to me that the clock shows times during the day well because they are single digit such as 8:30 but times in the pm would be double digit such as 20:22.


# TASK 5


## TASK 5 Part II
1. Display Cuts off the time shown when the time has more than 3 digits e.g. 11:15 shows 1:15
2. Provide user-friendly way to change color preferences

I came to this conclusion by analyzing the dominant topics in all the reviews as well as their ratings for my chosen assigned app. At least 6 out of 36(over 15%) of the reviews mention the fit of the time on the display for example REV-0 details that the border needs to be adjusted in order to display the full time which is bad for the overall user experience. Secondly, about another 5 reviews mention issues with being able to modify the font size and color which also has an adverse effect on the user experience.

## Bonus of TASK 5
1. https://github.com/wakhub/tinyclock/issues/3#issue-1197525051
2. https://github.com/wakhub/tinyclock/issues/4#issue-1197548619